Updated README primarily to cover modes of operation. (#196)
openshift/cloud-credential-operator: README.md

  * If the admin credentials fulfill neither of the above requirements, the controller will fail to generate the credentials, report failure back to the Cluster Version Operator, and thus block upgrading. The installer will also perform this check early to inform the user their cluster will not function.

## Cloud Providers

Currently the operator supports AWS, GCP, Azure, VMWare, OpenStack and oVirt.

## Admin Credential Locations

Cloud | Secret | Secret Data Keys
--- | --- | ---
AWS | kube-system/aws-creds | aws_access_key_id aws_secret_access_key
GCP | kube-system/gcp-credentials | service_account.json
Azure | kube-system/azure-credentials | azure_client_id azure_client_secret azure_subscription_id azure_tenant_id
OpenStack | kube-system/openstack-credentials | clouds.yaml
VSphere | kube-system/vsphere-creds |
oVirt | kube-system/ovirt-credentials | ovirt_url ovirt_username ovirt_password ovirt_cafile ovirt_insecure ovirt_ca_bundle

## Modes of Operation

### 1. Mint Mode

> The default and recommended best practice for running OpenShift is to run the installer with an admin level cloud credential. The admin credential is stored in kube-system namespace and then used by the cloud credential operator to process the CredentialRequests in the cluster and create additional fine grained permissions for each OpenShift component containing only the permissions it needs.

Reviewer: ```suggestion
The default and recommended best practice for running OpenShift is to run the installer with an admin level cloud credential. The admin credential is stored in kube-system namespace and then used by the cloud credential operator to process the CredentialRequests in the cluster and create new users for each CredentialsRequest with additional fine grained permissions for each OpenShift component containing only the permissions it needs.
```

Pros:
  * Each cluster component has only the permissions it needs.

>   * Automatic on-going reconciliation for cloud credentials including upgrades.

Reviewer: ```suggestion
  * Automatic on-going reconciliation for cloud credentials including upgrades (that require new/additional permissions).
```

Supported clouds: AWS, GCP, Azure

#### 1.1 Mint Mode With Removal/Rotation Of Admin Credential

Supported in OpenShift 4.5+

In this mode a user installs OpenShift with an admin credential per the normal mint mode, but removes the admin credential Secret from the cluster post-installation. The cloud credential operator makes it's own request for a read-only credential that allows it to verify if all CredentialsRequests have their required permissions, thus the admin credential is not needed unless something needs to be changed. (i.e. upgrade) Once removed the associated credential could then be destroyed on the underlying cloud if desired. However, this secret should be recreated prior to any upgrade and could then be removed again afterwards.

TODO: Credentials operator should report upgradable=false until the admin cred is restored.

TODO: document replacing the admin credential

Pros:
  * Admin credential is not stored in the cluster permanently and does not need to be long lived.
Cons:
  * Still requires admin credential in the cluster for brief periods of time.
  * Requires manually re-instating the Secret with admin credentials for each upgrade.

### 2. Passthrough Mode

> In this mode a user installs OpenShift with a single credential that *is not* an admin and cannot mint additional credentials, but itself has enough permissions to perform all operations needed by all components in the cluster. The cloud credential operator then shares this credential to each component.

Reviewer: ```suggestion
In this mode a user installs OpenShift with a single credential that *is not* an admin and cannot mint additional credentials, but itself has enough permissions to perform the installation and all operations needed by all components in the cluster. The cloud credential operator then shares this credential to each component.
```

Your passthrough mode credential will need to be manually maintained if CredentialsRequests change over time as the cluster is upgraded. This should be performed prior to every upgrade.

By default the permissions needed only for installation are required, however it is possible to reduce the permissions on your credential post-install to just what is needed to run the cluster. (as defined by the CredentialsRequests in the current release image)

TODO: treat like manual mode, require opt-in to upgrades or a preflight check

TODO: document rotating your passthrough credential.

TODO: document reducing permissions post install.

Pros:
  * Does not require installing or running with an admin credential.
Cons:
  * Includes broad permissions only needed at install time, unless manual action is taken to reduce permissions post-install.
  * Credential permissions may need to be manually updated prior to any upgrade.
  * Each component has permissions used by all other components.

Supported clouds: AWS, GCP, Azure, VMWare, OpenStack, oVirt

### 3. Manual Credentials Management

> In this mode a user manually performs the job of the cloud credential operator. This requires examinings the CredentialsRequests in an OpenShift 4 release image, creating credentials in the underlying cloud provider, and finally creating Kubernetes Secrets in the correct namespaces to satisfy

Reviewer: ```suggestion
In this mode a user manually performs the job of the cloud credential operator. This requires examining the CredentialsRequests in an OpenShift 4 release image, creating credentials in the underlying cloud provider, and finally creating Kubernetes Secrets in the correct namespaces to satisfy all CredentialsRequests for the current cloud.
```
